### DISCUSSION TITLE
[Proposal] Create innate-property-delegates.md #6204

### DISCUSSION THREAD

**Discussion by VWilcox2000** [dotnet/csharplang#6205] [Language Ideas]
*** Please skip to brief description below of the feature.  I misread the readme instructions and linked a pull (did not know what a git pull was, frankly at the start).  My apologies and thank you to MgSam and another MS rep for instruction on proper way to post. ***

**Comment by MgSam** [comment]
Could you post your suggestion in this discussion itself? Linking to a closed PR is not a great way for people to see what you're proposing. Also, maybe you could provide a terse example? I skimmed through your document but had a hard time understanding what you were actually proposing.

**Reply by VWilcox2000** [reply]
Thank you, Sam!  I appreciate the instruction -- I will post another comment focused on a summary.

**Comment by VWilcox2000** [comment]
**Brief Description of Innate-Property-Delegates (Bindable Property?)**

What I'm asking to consider is a property with built-in events and subscription options for initialization, retrieval, changing, changed, and possibly disposed.  This would shrink code for a number of very common use cases.  Additionally, it would be nice if there was an initialization option that would allow 'this' for in-line construction similar to the lazy initialization I believe is slated for C# 11.

The bindable property could be identified with a keyword or a symbol such as property, bindable, or a symbol like this:

public **property** int age;
public **bindable** int age;
public int^ age;

Used in a sample class below below:

      class SomeClass
      {
        public property int age;  // simple but maybe confusing as int age { get; set; } is also a property
        public bindable int age;
        public int^ age;          // or some other suitable symbol not reserved.

        //This would allow the user to wire events with += or subscription pattern.
        public SomeClass()
        {
          this.age = 10;
          this.age.changing += this.onAgeChanging;
          this.age.changed += this.onAgeChanged;
        }

        private void onAgeChanging(object sender, ChangingEventArgs<int> args)
        {
          if (args.newValue < 1 || args.newValue > 100)
          {
            args.preventChange = true;
          }
        }

        private void onAgeChanged(object sender, ChangedEventArgs<int> args)
        {
          if (args.newValue < 18)
          {
            // hide all but children friendly content
          }
        }

In the sample, I am omitting { get; set; } as the marking already informs compiler this is a property, but they could be allowed.

      class SomeClass
      {
        public bindable int age
        {
          get
          {
            // on retrieving event called first
            return Math.Min(this.ageMax, value);  // <-- value is the value of the field -- what the int age literally is
          }
          set
          {
            // on changing called
            if (change.prevented)   < --change innate keyword like value --with the ChaningEventArgs<i>
            {
              // do something

            }
            // on change called
          }
        }

Finally, I would want the ability to inline -- initialize the property.

        class SomeClass
        {
          public bindable int age = 8;  // simple enough
          public bindable PersonData person = new PersonData();  // simple, but what if PersonData needs a ref to us
          public bindable PersonData person ==> new PersonData(this); // ==> or some other suitable syntax similar to the late binding I believe with ?? = would make this
                                                                      // kind of inline construction possible.  this assignment should happen either at the beginning of the constructor or end (I'd think beginning) where this is available.

Why I would like this change is because it simplifies very frequently used code patterns by many.

A sample use in **MAUI** which is one of the places I think this feature would benefit the most.

        public sealed class MyViewModel : MAUIModel
        {
          public bindable int age = 6;
          public bindable string? firstName;
          public bindable string? lastName;

          public Command register = new Command(...)


Now, in my view, I'll use XAML, I can do the following:

        <Entry Text="{Binding age}" />
        <Entry Text="{Binding firstName}" />
        <Entry Text="{Binding lastName}" />

For non-Xamarin / non-MAUI programmers, to make this happen requires a fair amount of code for each property as follows:

        public sealed class MyViewModel : INotifyPropertyChanged
        {
          public event PropertyChangedEventHandler? PropertyChanged;

          private string? _firstName = null;
          public string? firstName
          {
            get => this._firstName;
            set
            {
              if (this._firstName != value)
              {
                this._firstName = value;
                this.PropertyChanged?.Invoke(FirstNameProperty.PropertyName);
              }
            }
          }

          public static readonly BindableProperty FirstNameProperty =
            BindableProperty.Create(
              nameof(firstName),
              typeof(string),
              typeof(MyViewModel),
              null);

          // the same needed again for age, lastName
        }


**Entity Framework** change tracking might benefit from the same functionality.  Rather than comparing original records to current state to detect changes -- by subscribing to changes fields in a property it would keep track of changed fields and leverage that during SaveChanges.



Another use is in **data synchronization**:
One of my main Blazor projects coordinates multiple users connected to a single document using a synchronization system.  To achieve this, my models all leverage a base class that maps all items, and runs synchronization -- in addition, all of my data is wrapped in a class:

      public sealed class SyncModel : SyncData
      {
        public SessionValue<int> age { get; } = new SessionValue<int>();

        private void SomeCode()
        {
          this.age.value = 8;
          this.age.valueChanged += this.onAgeChanged(...


This does for all my data what ObservableCollection does to collections(though far simple on the event side).

The quality of life of such systems would be greatly increased with replacing SessionValue<int> with bindable int age;

Having the option to wire the events via reflection is very important to me as I use a significant amount of reflection based automation to coordinate models and synchronization.  For reflection, a IBindablePropertyInfo could be added, or an optional IBindable? property could be added to IPropertyInfo.  I map classes and discover both what I call simple items and synchronizable items, and wire events automatically to make some magic happen.  With this feature, I could do that without replacing every int, float, string, etc. with a helper class that holds them.

In similar fashion to how nullable values use INullable, having a system interface for the bindable properties would likely be the easiest way to interact with them.  A quick idea:

      public interface IBindableProperty<T> : IStructuralEquatable<T>
      {
        event InitializedValueHandler<T>? initialized;
        event ChangingValueHandler<T>? changing;
        event ChangedValueHandler<T>? changed;
        event RetrievingValueHandler<T>? retrieving;

        // consider subscription model 
        void onInitialized(Action<InitializedValueHandler<T>> handler) ...
          ... etc.
      }

The point of IStructuralEquatable<T> is that if a user wants to use their own classes as bindable properties, they MUST support that (or optionally could allow IComparable) so that the system can detect changes.

Another possibility would be allowing inline handling rather than wiring or subscribing to events with handlers near the property similar to the get / set structure with something vaguely like this:

      public bindable PersonData person ==> new PersonData(this)
        changing => this.ValidateUser(changing),
        changed => this.InvokeAsync(this.StateHasChanged);
